template.rb

```diff
@@ -1,9 +1,69 @@
 require 'shellwords'
 
+# Set Thor::Actions source path for looking up the files
+def source_paths
+  @source_paths
+end
+
+# Prepend the required paths to the source paths to make the template files in those paths available
+def use_source_path(source_path)
+  @source_paths.unshift(source_path)
+end
+
+def remote_repository
+  require 'tmpdir'
+  tempdir = Dir.mktmpdir('rails-templates')
+  at_exit { FileUtils.remove_entry(tempdir) }
+
+  git clone: [
+    '--quiet',
+    'https://github.com/nimblehq/rails-templates.git',
+    tempdir
+  ].map(&:shellescape).join(' ')
+
+  if (branch = __FILE__[%r{rails-templates/(.+)/template.rb}, 1])
+    Dir.chdir(tempdir) { git checkout: branch }
+  end
+
+  tempdir
+end
+
+def delete_test_folder
+  FileUtils.rm_rf('test')
+end
+
+def print_error_message
+  puts <<~EOT
+    #{'=' * 80}
+
+    There are some errors when templating the application, Please fix them manually:
+
+    #{@template_errors}
+    #{'=' * 80}
+  EOT
+end
+
+# Setup the template root path
+# If the template file is the url, clone the repo to the tmp directory
+template_root = __FILE__ =~ %r{\Ahttps?://} ? remote_repository : __dir__
+
+# Init the template lib
+require "#{template_root}/.template/lib/template/application"
+require "#{template_root}/.template/lib/template/errors"
+
+@template_application = Template::Application.new(app_name)
+@template_errors = Template::Errors.new
+
+# Init the source path
+@source_paths ||= []
+use_source_path template_root
+
 # Variables
```

Comment:
The variables are placed too far down now since lots of methods have been pasted above 🤔 It would be better to keep those at the top of the file.

Reply:
Yeah It feels quite weird to put here. The reason is it need to init from the class we required above. So we need to clare it before declaring the variables

APP_NAME_HUMANIZED =  `@template_application`.humanized_name

-----

I think I'm going to call `@template_application.humanized_name` directly instead of stored it in the variables to avoid this weird ordering 👍